Mathlib/Tactic/Widget/CommDiag.lean

```diff
@@ -0,0 +1,137 @@
+/-
+Copyright (c) 2022 Wojciech Nawrocki. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Wojciech Nawrocki
+-/
+
+import Mathlib.CategoryTheory.Category.Basic
+
+import ProofWidgets.Component.PenroseDiagram
+import ProofWidgets.Presentation.Expr
+import ProofWidgets.Component.SelectionPanel
+
+/-! This module defines tactic/meta infrastructure for displaying commutative diagrams in the
+infoview. -/
+
+open Lean in
+/-- If the expression is a function application of `fName` with 7 arguments, return those arguments.
+Otherwise return `none`. -/
+@[inline] def _root_.Lean.Expr.app7? (e : Expr) (fName : Name)
+    : Option (Expr × Expr × Expr × Expr × Expr × Expr × Expr) :=
+  if e.isAppOfArity fName 7 then
+    some (
+      e.appFn!.appFn!.appFn!.appFn!.appFn!.appFn!.appArg!,
+      e.appFn!.appFn!.appFn!.appFn!.appFn!.appArg!,
+      e.appFn!.appFn!.appFn!.appFn!.appArg!,
+      e.appFn!.appFn!.appFn!.appArg!,
+      e.appFn!.appFn!.appArg!,
+      e.appFn!.appArg!,
+      e.appArg!
+    )
+  else
+    none
+
+namespace Mathlib.Tactic.Widget
+open Lean Meta
+open ProofWidgets
+open CategoryTheory
+
+/-! ## Metaprogramming utilities for breaking down category theory expressions -/
+
+/-- Given a Hom type `α ⟶ β`, return `(α, β)`. Otherwise `none`. -/
+def homType? (e : Expr) : Option (Expr × Expr) := do
+  let some (_, _, A, B) := e.app4? ``Quiver.Hom | none
+  return (A, B)
+
+/-- Given composed homs `g ≫ h`, return `(g, h)`. Otherwise `none`. -/
+def homComp? (f : Expr) : Option (Expr × Expr) := do
+  let some (_, _, _, _, _, f, g) := f.app7? ``CategoryStruct.comp | none
+  return (f, g)
+
+/-- Expressions to display as labels in a diagram. -/
+abbrev ExprEmbeds := Array (String × Expr)
+
+/-! ## Widget for general commutative diagrams -/
+
+open scoped Jsx in
+/-- Construct a commutative diagram from a Penrose `sub`stance program and expressions `embeds` to
+display as labels in the diagram. -/
+def mkCommDiag (sub : String) (embeds : ExprEmbeds) : MetaM Html := do
+  let embeds ← embeds.mapM fun (s, h) =>
+      return (s, Html.ofTHtml <InteractiveCode fmt={← Widget.ppExprTagged h} />)
+  return Html.ofTHtml
+    <PenroseDiagram
+      embeds={embeds}
+      dsl={include_str ".."/".."/".."/"widget"/"src"/"penrose"/"commutative.dsl"}
+      sty={include_str ".."/".."/".."/"widget"/"src"/"penrose"/"commutativeOpt.sty"}
+      sub={sub} />
+
+/-! ## Commutative triangles -/
+
+/--
+Triangle with `homs = [f,g,h]` and `objs = [A,B,C]`
+```
+A f B
+  h g
+    C
+``` -/
```

Comment:
This seems to have broken the syntax higlighter. I think
````suggestion
/--
Triangle with `homs = [f,g,h]` and `objs = [A,B,C]`
```
A f B
  h g
    C
```
-/
````
fixes it